From 59f87b7556bb4be028524853304f9e4a56d75cb5 Mon Sep 17 00:00:00 2001
From: Ryan Gonzalez <git@refi64.dev>
Date: Mon, 24 Apr 2023 04:22:10 +0000
Subject: [PATCH] gl: Ensure EGL errors are preserved

eglGetError() clear the current error, which means that the after the
check for EGL_BAD_MATCH is performed, LOG(ERROR) can't access the
original error anymore. Make sure the error gets saved and then reused
when logging at the end.

Bug: None
Change-Id: Idb8a3fe95caeadcab374f8694158a740d378b118
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4461396
Commit-Queue: Peng Huang <penghuang@chromium.org>
Reviewed-by: Peng Huang <penghuang@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1134390}
---
 ui/gl/gl_context_egl.cc | 22 ++++++++++++----------
 1 file changed, 12 insertions(+), 10 deletions(-)

diff --git a/ui/gl/gl_context_egl.cc b/ui/gl/gl_context_egl.cc
index bc10dedbf51c3..08f44200745a4 100644
--- a/ui/gl/gl_context_egl.cc
+++ b/ui/gl/gl_context_egl.cc
@@ -90,6 +90,7 @@
 #define EGL_CONTEXT_VIRTUALIZATION_GROUP_ANGLE 0x3481
 #endif /* EGL_ANGLE_context_virtualization */
 
+using ui::GetEGLErrorString;
 using ui::GetLastEGLErrorString;
 
 namespace gl {
@@ -343,12 +344,15 @@ bool GLContextEGL::Initialize(GLSurface* compatible_surface,
       eglCreateContext(gl_display_->GetDisplay(), config_,
                        share_group() ? share_group()->GetHandle() : nullptr,
                        context_attributes.data());
+  if (context_) {
+    return true;
+  }
 
   // If EGL_KHR_no_config_context is in use and context creation failed,
   // it might indicate that an unsupported ES version was requested. Try
   // falling back to a lower version.
-  if (!context_ && gl_display_->ext->b_EGL_KHR_no_config_context &&
-      eglGetError() == EGL_BAD_MATCH) {
+  GLint error = eglGetError();
+  if (gl_display_->ext->b_EGL_KHR_no_config_context && error == EGL_BAD_MATCH) {
     // Set up the list of versions to try: 3.1 -> 3.0 -> 2.0
     std::vector<std::pair<EGLint, EGLint>> candidate_versions;
     if (context_client_major_version == 3 &&
@@ -374,18 +378,16 @@ bool GLContextEGL::Initialize(GLSurface* compatible_surface,
                            context_attributes.data());
       // Stop searching as soon as a context is successfully created.
       if (context_) {
-        break;
+        return true;
+      } else {
+        error = eglGetError();
       }
     }
   }
 
-  if (!context_) {
-    LOG(ERROR) << "eglCreateContext failed with error "
-               << GetLastEGLErrorString();
-    return false;
-  }
-
-  return true;
+  LOG(ERROR) << "eglCreateContext failed with error "
+             << GetEGLErrorString(error);
+  return false;
 }
 
 void GLContextEGL::Destroy() {
-- 
2.39.2

